onnxruntime/contrib_ops/cuda/bert/bert_padding.cu

```diff
@@ -0,0 +1,342 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+// TrtSequenceOffset kernels are modified from FasterTransformer
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "core/providers/cuda/cuda_common.h"
+#include "contrib_ops/cuda/bert/bert_padding.h"
+
+using namespace onnxruntime::cuda;
+
+namespace onnxruntime {
+namespace contrib {
+namespace cuda {
+
+constexpr int32_t kMAX_THREADS_PER_BLOCK = 256;
+
+// -----------------------------------
+// Get indices of non-padding tokens and padding tokens. Here we assume that padding is on the right side of sequence.
+// sequence_token_count is number of non-padding tokens per sequence, and it has shape [batch_size].
+// For example, we have 3 sequences with 1, 2, 4 non-padding tokens and positions like the following (* means padding):
+//   Sequence_0:   0,  1*, 2*,  3*
+//   Sequence_1:   4,  5,  6*,  7*
+//   Sequence_2:   8,  9,  10,  11
+// token_offset: 0, 4, 5, 8, 9, 10, 11,  1*, 2*, 3*, 6*, 7*
+// token_count_buffer has two numbers for non-padding tokens:
+//   total_token_count: 1 + 2 + 4 = 7
+//   max_token_count: 4
+// cumulated_token_count: 0, 1, 1+2, 1+2+4
+__global__ void getTokenOffset(int* token_count_buffer,
```

Comment:
>_global__ void getTokenOffset(int* token_count_buffer, [](http://example.com/codeflow?start=1&length=54)

It can be implemented with cub::BlockScan. The kernel can be launched with Grid: 1, Block: batch. 
For kernel: 
1. it uses cub::BlockScan to compute cumulated_token_count firstly. 
2. then each thread fills its token_offset

Reply:
> > _global__ void getTokenOffset(int* token_count_buffer,
> 
> It can be implemented with cub::BlockScan. The kernel can be launched with Grid: 1, Block: batch. For kernel:
> 
> 1. it uses cub::BlockScan to compute cumulated_token_count firstly.
> 2. then each thread fills its token_offset

Good suggestion. There is a TODO in comment that is related to this:
// TODO(tianleiwu): Use cub::DevicePartition::Flagged like BuildGlobalIndex in longformer_global_impl.cu
//                  to build token_offset when sequence length is large.
I could do it in another pull request later.

Reply:
> For offset idx > token_size, we don't need to fill it actually because the restoring won't use it

The purpose is to fill zeros for those padded tokens (to make result determined). Otherwise, we will need the fill the whole output with zeros first, then use another kernel to restore non-padding tokens.

Another purpose is to make the shape as (batch_size, sequence_length). Otherwise, we will need pass these two values  to restore padding operator.